File: platform-jvm/src/main/java/org/sweetchips/platform/jvm/JvmContext.java
package org.sweetchips.platform.jvm;

import org.objectweb.asm.ClassReader;
import org.objectweb.asm.ClassVisitor;
import org.objectweb.asm.ClassWriter;
import org.objectweb.asm.tree.ClassNode;
import org.sweetchips.platform.common.ContextLogger;
import org.sweetchips.platform.common.PlatformContext;
import org.sweetchips.utility.AsyncUtil;
import org.sweetchips.utility.ItemsUtil;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Deque;
import java.util.List;
import java.util.Map;
import java.util.Queue;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentLinkedQueue;
import java.util.function.BiConsumer;
import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.Supplier;

public final class JvmContext implements PlatformContext {

    private static final String TAG = "JvmContext";

    private final ContextLogger mLogger;

    public JvmContext(ContextLogger logger) {
        mLogger = logger;
    }

    private boolean mIncremental;
    private int mApi;
    private Map<Object, Object> mExtra = new ConcurrentHashMap<>();
    private Deque<ClassVisitorFactory> mPrepare = new ArrayDeque<>();
    private Deque<ClassVisitorFactory> mTransform = new ArrayDeque<>();
    private Collection<Supplier<ClassNode>> mAdditions = new ConcurrentLinkedQueue<>();
    private List<Consumer<Map<Object, Object>>> mPrepareBefore = new ArrayList<>();
    private List<Consumer<Map<Object, Object>>> mPrepareAfter = new ArrayList<>();
    private List<Consumer<Map<Object, Object>>> mTransformBefore = new ArrayList<>();
    private List<Consumer<Map<Object, Object>>> mTransformAfter = new ArrayList<>();
    private Collection<ClassNode> mClasses;
    private BiConsumer<String, byte[]> mBytesWriter;

    @Override
    public Runnable onPrepareBefore() {
        return this::doPrepareBefore;
    }

    public Consumer<byte[]> onPrepare() {
        return this::doPrepare;
    }

    @Override
    public Runnable onPrepareAfter() {
        return this::doPrepareAfter;
    }

    @Override
    public Runnable onTransformBefore() {
        return this::doTransformBefore;
    }

    public Function<byte[], byte[]> onTransform() {
        return this::doTransform;
    }

    @Override
    public Runnable onTransformAfter() {
        return this::doTransformAfter;
    }

    public boolean isIncremental() {
        return mIncremental;
    }

    public void setIncremental(boolean incremental) {
        mIncremental = incremental;
    }

    public void setApi(int api) {
        mApi = api;
    }

    public int getApi() {
        return mApi;
    }

    public void setBytesWriter(BiConsumer<String, byte[]> bytesWriter) {
        mBytesWriter = bytesWriter;
    }

    public void setExtra(Map<Object, Object> extra) {
        mExtra = extra;
    }

    public Map<Object, Object> getExtra() {
        return mExtra;
    }

    public void addPrepareBefore(Consumer<Map<Object, Object>> consumer) {
        ItemsUtil.checkAndAdd(mPrepareBefore, consumer);
    }

    public void addPrepareFirst(ClassVisitorFactory factory) {
        ItemsUtil.checkAndAddFirst(mPrepare, factory);
    }

    public void addPrepareLast(ClassVisitorFactory factory) {
        ItemsUtil.checkAndAdd(mPrepare, factory);
    }

    public void addPrepareAfter(Consumer<Map<Object, Object>> consumer) {
        ItemsUtil.checkAndAdd(mPrepareAfter, consumer);
    }

    public void addClass(Supplier<ClassNode> supplier) {
        ItemsUtil.checkAndAdd(mAdditions, supplier);
    }

    public void addTransformBefore(Consumer<Map<Object, Object>> consumer) {
        ItemsUtil.checkAndAdd(mTransformBefore, consumer);
    }

    public void addTransformFirst(ClassVisitorFactory factory) {
        ItemsUtil.checkAndAddFirst(mTransform, factory);
    }

    public void addTransformLast(ClassVisitorFactory factory) {
        ItemsUtil.checkAndAdd(mTransform, factory);
    }

    public void addTransformAfter(Consumer<Map<Object, Object>> consumer) {
        ItemsUtil.checkAndAdd(mTransformAfter, consumer);
    }

    private void doPrepareBefore() {
        mLogger.d(TAG, "doPrepareBefore: begin");
        AsyncUtil.with(mPrepareBefore.stream()).forkJoin(it -> it.accept(mExtra));
        mPrepareBefore = null;
        mLogger.d(TAG, "doPrepareBefore: end");
    }

    private void doPrepare(byte[] bytes) {
        Collection<ClassVisitorFactory> collection = mPrepare;
        if (collection.size() == 0) {
            return;
        }
        ClassReader cr = new ClassReader(bytes);
        ClassVisitor cv = null;
        for (ClassVisitorFactory factory : collection) {
            cv = factory.newInstance(mApi, cv, mExtra);
        }
        cr.accept(cv, ClassReader.EXPAND_FRAMES);
    }

    private void doPrepareAdditions() {
        mLogger.d(TAG, "doPrepareAdditions: begin");
        Collection<ClassVisitorFactory> collection = mPrepare;
        mPrepare = null;
        Queue<ClassNode> queue = new ConcurrentLinkedQueue<>();
        mClasses = queue;
        AsyncUtil.with(mAdditions.stream()).forkJoin(it -> {
            ClassNode addition = it.get();
            mLogger.i(TAG, "doPrepareAdditions: create " + addition.name);
            if (collection.size() == 0) {
                queue.add(addition);
                return;
            }
            ClassNode cn = new ClassNode(mApi);
            ClassVisitor cv = cn;
            for (ClassVisitorFactory factory : collection) {
                cv = factory.newInstance(mApi, cv, mExtra);
            }
            addition.accept(cv);
            queue.add(cn);
        });
        mAdditions = null;
        mLogger.d(TAG, "doPrepareAdditions: end");
    }

    private void doPrepareAfter() {
        doPrepareAdditions();
        mLogger.d(TAG, "doPrepareAfter: begin");
        AsyncUtil.with(mPrepareAfter.stream()).forkJoin(it -> it.accept(mExtra));
        mPrepareAfter = null;
        mLogger.d(TAG, "doPrepareAfter: end");
    }

    private void doTransformBefore() {
        mLogger.d(TAG, "doTransformBefore: begin");
        AsyncUtil.with(mTransformBefore.stream()).forkJoin(it -> it.accept(mExtra));
        mTransformBefore = null;
        mLogger.d(TAG, "doTransformBefore: end");
        doTransformAdditions();
    }

    private void doTransformAdditions() {
        mLogger.d(TAG, "doTransformAdditions: begin");
        Collection<ClassVisitorFactory> collection = mTransform;
        Collection<ClassNode> classes = mClasses;
        mClasses = null;
        AsyncUtil.with(classes.stream()).forkJoin(it -> {
            if (collection.size() == 0) {
                checkAndWriteBytes(it);
                return;
            }
            ClassNode cn = new ClassNode(mApi);
            ClassVisitor cv = cn;
            for (ClassVisitorFactory factory : collection) {
                cv = factory.newInstance(mApi, cv, mExtra);
            }
            ClassesSetting.resetDeleteFlag();
            it.accept(cv);
            if (ClassesSetting.checkDeleteFlag()) {
                mLogger.i(TAG, "doTransformAdditions: delete " + cn.name);
                return;
            }
            checkAndWriteBytes(cn);
        });
        mBytesWriter = null;
        mLogger.d(TAG, "doTransformAdditions: end");
    }

    private byte[] doTransform(byte[] bytes) {
        Collection<ClassVisitorFactory> collection = mTransform;
        if (collection.size() == 0) {
            return bytes;
        }
        ClassReader cr = new ClassReader(bytes);
        ClassWriter cw = new ClassWriter(mApi);
        ClassVisitor cv = cw;
        for (ClassVisitorFactory factory : collection) {
            cv = factory.newInstance(mApi, cv, mExtra);
        }
        ClassesSetting.resetDeleteFlag();
        cr.accept(cv, ClassReader.EXPAND_FRAMES);
        if (ClassesSetting.checkDeleteFlag()) {
            mLogger.i(TAG, "doTransform: delete a class");
            return null;
        }
        return cw.toByteArray();
    }

    private void doTransformAfter() {
        mLogger.d(TAG, "doTransformAfter: begin");
        mTransform = null;
        AsyncUtil.with(mTransformAfter.stream()).forkJoin(it -> it.accept(mExtra));
        mTransformAfter = null;
        mExtra = null;
        mLogger.d(TAG, "doTransformAfter: end");
    }

    private void checkAndWriteBytes(ClassNode cn) {
        BiConsumer<String, byte[]> bytesWriter = mBytesWriter;
        if (bytesWriter != null) {
            String name = cn.name;
            ClassWriter cw = new ClassWriter(mApi);
            cn.accept(cw);
            bytesWriter.accept(name, cw.toByteArray());
        }
    }
}

File: platform-jvm/src/main/kotlin/org/sweetchips/platform/jvm/JvmContext.kt
package org.sweetchips.platform.jvm

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withContext
import org.objectweb.asm.ClassReader
import org.objectweb.asm.ClassVisitor
import org.objectweb.asm.ClassWriter
import org.objectweb.asm.tree.ClassNode
import org.sweetchips.platform.common.ContextLogger
import org.sweetchips.platform.common.PlatformContext
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ConcurrentLinkedQueue
import java.util.function.BiConsumer
import java.util.function.Consumer
import java.util.function.Function
import java.util.function.Supplier

class JvmContext(val logger: ContextLogger) : PlatformContext {

    private companion object {
        const val TAG = "JvmContext"
    }

    var api: Int = 0
    var isIncremental: Boolean = false
    var extra: MutableMap<Any, Any>? = ConcurrentHashMap()

    private var bytesWriter: BiConsumer<String, ByteArray>? = null

    private var prepare: Deque<ClassVisitorFactory>? = ArrayDeque()
    private var transform: Deque<ClassVisitorFactory>? = ArrayDeque()

    private var prepareBefore: MutableList<Consumer<Map<Any, Any>>>? = mutableListOf()
    private var prepareAfter: MutableList<Consumer<Map<Any, Any>>>? = mutableListOf()
    private var transformBefore: MutableList<Consumer<Map<Any, Any>>>? = mutableListOf()
    private var transformAfter: MutableList<Consumer<Map<Any, Any>>>? = mutableListOf()

    private var additions: MutableCollection<Supplier<ClassNode>>? = ConcurrentLinkedQueue()
    private var classes: MutableCollection<ClassNode>? = null

    fun setBytesWriter(consumer: BiConsumer<String, ByteArray>) {
        bytesWriter = consumer
    }

    fun addPrepareBefore(action: Consumer<Map<Any, Any>>) {
        prepareBefore?.add(action)
    }

    fun addPrepareAfter(action: Consumer<Map<Any, Any>>) {
        prepareAfter?.add(action)
    }

    fun addTransformBefore(action: Consumer<Map<Any, Any>>) {
        transformBefore?.add(action)
    }

    fun addTransformAfter(action: Consumer<Map<Any, Any>>) {
        transformAfter?.add(action)
    }

    fun addPrepareFirst(action: ClassVisitorFactory?) {
        prepare?.addFirst(action)
    }

    fun addPrepareLast(action: ClassVisitorFactory?) {
        prepare?.addLast(action)
    }

    fun addTransformFirst(action: ClassVisitorFactory?) {
        transform?.addFirst(action)
    }

    fun addTransformLast(action: ClassVisitorFactory?) {
        transform?.addLast(action)
    }

    fun addClass(action: Supplier<ClassNode>) {
        additions?.add(action)
    }

    override fun onPrepareBefore() = Runnable {
        prepareBefore!!.also { prepareBefore = null }.phase("prepareBefore")
    }

    override fun onPrepareAfter() = Runnable {
        onPrepareAdditions()
        prepareAfter!!.also { prepareAfter = null }.phase("prepareAfter")
    }

    override fun onTransformBefore() = Runnable {
        transformBefore!!.also { transformBefore = null }.phase("transformBefore")
        onTransformAdditions()
    }

    override fun onTransformAfter() = Runnable {
        transform = null
        transformAfter!!.also { transformAfter = null }.phase("transformAfter")
        extra = null
    }

    fun onPrepare() = Consumer<ByteArray> { bytes ->
        val actions = prepare!!
        if (actions.size == 0) {
            return@Consumer
        }
        val cr = ClassReader(bytes)
        var cv = object : ClassVisitor(api) {} as ClassVisitor
        actions.forEach {
            cv = it.newInstance(api, cv, extra)
        }
        cr.accept(cv, ClassReader.EXPAND_FRAMES)
    }

    fun onTransform() = Function<ByteArray, ByteArray?> { bytes ->
        val actions = transform!!
        if (actions.size == 0) {
            return@Function bytes
        }
        val cr = ClassReader(bytes)
        val cw = ClassWriter(api)
        var cv = cw as ClassVisitor
        actions.forEach {
            cv = it.newInstance(api, cv, extra)
        }
        cr.accept(cv, ClassReader.EXPAND_FRAMES)
        if (ClassesSetting.checkDeleteFlag()) {
            val tag = "onTransform"
            logger.i(TAG, "$tag: delete a class")
            return@Function null
        }
        return@Function cw.toByteArray()
    }

    private fun List<Consumer<Map<Any, Any>>>.phase(tag: String) {
        logger.d(TAG, "$tag: begin")
        forEach {
            it.accept(extra!!)
        }
        logger.d(TAG, "$tag: end")
    }

    private fun onPrepareAdditions() {
        val tag = "onPrepareAdditions"
        logger.d(TAG, "$tag: begin")
        val actions = prepare!!.also { prepare = null }
        val nodes = ConcurrentLinkedQueue<ClassNode>().also { classes = it }
        runBlocking {
            additions!!.also { additions = null }.forEach {
                launch {
                    val node = withContext(Dispatchers.IO) { it.get() }
                    logger.i(TAG, "$tag: create ${node.name}")
                    if (actions.size == 0) {
                        nodes.add(node)
                        return@launch
                    }
                    val cn = ClassNode(api)
                    var cv = cn as ClassVisitor
                    actions.forEach {
                        cv = it.newInstance(api, cv, extra)
                    }
                    withContext(Dispatchers.Default) {
                        node.accept(cv)
                    }
                    nodes.add(cn)
                }
            }
        }
        logger.d(TAG, "$tag: end")
    }

    private fun onTransformAdditions() {
        val tag = "onTransformAdditions"
        logger.d(TAG, "$tag: begin")
        val actions = transform!!
        val writeNode = bytesWriter?.also { bytesWriter = null }?.run {
            { cn: ClassNode ->
                val name = cn.name
                val cw = ClassWriter(api)
                cn.accept(cw)
                accept(name, cw.toByteArray())
            }
        } ?: {
            Unit
        }
        runBlocking {
            classes!!.also { classes = null }.forEach {
                launch {
                    if (actions.size == 0) {
                        writeNode(it)
                        return@launch
                    }
                    val cn = ClassNode(api)
                    var cv = cn as ClassVisitor
                    actions.forEach {
                        cv = it.newInstance(api, cv, extra)
                    }
                    withContext(Dispatchers.Default) {
                        it.accept(cv)
                    }
                    if (ClassesSetting.checkDeleteFlag()) {
                        logger.i(TAG, "$tag: delete ${cn.name}")
                        return@launch
                    }
                    writeNode(cn)
                }
            }
        }
        logger.d(TAG, "$tag: end")
    }
}